framework/src/main/java/org/checkerframework/framework/source/CompositeChecker.java

```diff
@@ -0,0 +1,684 @@
+    }
+
+    return Collections.unmodifiableList(immediateSubcheckers);
+  }
+
+  /**
+   * Get the list of all subcheckers (if any). via the instantiateSubcheckers method. This list is
+   * only non-empty for the one checker that runs all other subcheckers. These are recursively
+   * instantiated via instantiateSubcheckers the first time the method is called if subcheckers is
+   * null. Assumes all checkers run on the same thread.
+   *
+   * @return the list of all subcheckers (if any)
+   */
+  public List<BaseTypeChecker> getSubcheckers() {
+    if (subcheckers == null) {
+      // Instantiate the checkers this one depends on, if any.
+      Map<Class<? extends BaseTypeChecker>, BaseTypeChecker> checkerMap =
+          new ArrayMap<Class<? extends BaseTypeChecker>, BaseTypeChecker>(2);
+
+      immediateSubcheckers = instantiateSubcheckers(checkerMap);
+
+      subcheckers = Collections.unmodifiableList(new ArrayList<>(checkerMap.values()));
+    }
+
+    return subcheckers;
+  }
+
+  /**
+   * Get the shared TreePathCacher instance.
+   *
+   * @return the shared TreePathCacher instance.
+   */
+  public TreePathCacher getTreePathCacher() {
+    if (treePathCacher == null) {
+      // In case it wasn't already set in instantiateSubcheckers.
+      treePathCacher = new TreePathCacher();
+    }
+    return treePathCacher;
+  }
+
+  // AbstractTypeProcessor delegation
+  @Override
+  public void typeProcess(TypeElement element, TreePath tree) {
+    if (!getSubcheckers().isEmpty()) {
+      // TODO: I expected this to only be necessary if (parentChecker == null).
+      // However, the NestedAggregateChecker fails otherwise.
+      messageStore.clear();
+    }
+
+    // Errors (or other messages) issued via
+    //   SourceChecker#message(Diagnostic.Kind, Object, String, Object...)
+    // are stored in messageStore until all checkers have processed this compilation unit.
+    // All other messages are printed immediately.  This includes errors issued because the
+    // checker threw an exception.
+
+    // In order to run the next checker on this compilation unit even if the previous issued
+    // errors, the next checker's errsOnLastExit needs to include all errors issued by previous
+    // checkers.
+
+    Context context = ((JavacProcessingEnvironment) processingEnv).getContext();
+    Log log = Log.instance(context);
+
+    int nerrorsOfAllPreviousCheckers = this.errsOnLastExit;
+    for (BaseTypeChecker subchecker : getSubcheckers()) {
+      subchecker.errsOnLastExit = nerrorsOfAllPreviousCheckers;
+      subchecker.messageStore = messageStore;
+      int errorsBeforeTypeChecking = log.nerrors;
+
+      subchecker.typeProcess(element, tree);
+
+      int errorsAfterTypeChecking = log.nerrors;
+      nerrorsOfAllPreviousCheckers += errorsAfterTypeChecking - errorsBeforeTypeChecking;
+    }
+
+    this.errsOnLastExit = nerrorsOfAllPreviousCheckers;
+    super.typeProcess(element, tree);
```

Comment:
Why does this call `super.typeProcess`, as opposed to `this.typeProcess`?

Reply:
That would be a recursive call (it's in the typeProcess(TypeElement, TreePath) method).  Looks to me like just a normal super call in a method override.

Also, I didn't write this, it's just moved up into the CompositeChecker from the BaseTypeChecker. It's not new code.

Reply:
Thank you for the review! Suzanne is already working on a new PR to clean the BaseTypeChecker refactor up.

Reply:
Not sure how I missed it would be recursive! My mistake.